MM-62960: Improve webConn remove performance from hubConnectionIndex

agnivade opened the pull request:
When we added iteration by channelID, this was a known tradeoff during that.
However, it has been observed that the regular connection removal function
creates considerable blocking of the processing loop, leading to high
CPU usage and API latencies.

To fix this, we add a reverse mapping of channelIDs to connections
and their positions in the slice. This helps us to remove the connection
from the slice without iteration.

Unfortunately, this still needs to iterate through all channelIDs
during invalidation of the channel member cache. However, the user
cache invalidation is not a regular activity. So it should be an acceptable
tradeoff to make.

https://mattermost.atlassian.net/browse/MM-62960

```release-note
NONE
```

agnivade commented:
I have manually tested an improvement with this test case:

```go
 func TestHubConnIndexRemove(t *testing.T) {
	th := Setup(t).InitBasic()
	defer th.TearDown()
	connIndex := newHubConnectionIndex(1*time.Second, th.Service.Store, th.Service.logger)

	// Setup: Create many users and channels
	numChannels := 100
	numUsers := 10
	// Create some users
	userIds := make([]string, numUsers)
	for i := range userIds {
		userIds[i] = th.CreateUserOrGuest(false).Id
	}

	// Create channels and add most users to most channels
	channelIds := make([]string, numChannels)
	for i := 0; i < numChannels; i++ {
		ch := th.CreateChannel(th.BasicTeam)
		channelIds[i] = ch.Id

		// Add most users to this channel
		_, err := th.Service.Store.Channel().SaveMember(th.Context, &model.ChannelMember{
			ChannelId:   ch.Id,
			UserId:      th.BasicUser.Id,
			NotifyProps: model.GetDefaultChannelNotifyProps(),
		})
		require.NoError(t, err)

		for _, userID := range userIds {
			_, err = th.Service.Store.Channel().SaveMember(th.Context, &model.ChannelMember{
				ChannelId:   ch.Id,
				UserId:      userID,
				NotifyProps: model.GetDefaultChannelNotifyProps(),
			})
			require.NoError(t, err)

		}
	}

	// Create a special user who is only in 2 channels
	specialUserID := th.CreateUserOrGuest(false).Id
	for i := 0; i < 2; i++ {
		_, err := th.Service.Store.Channel().SaveMember(th.Context, &model.ChannelMember{
			ChannelId:   channelIds[i],
			UserId:      specialUserID,
			NotifyProps: model.GetDefaultChannelNotifyProps(),
		})
		require.NoError(t, err)
	}

	wc := &WebConn{
		Platform: th.Service,
		Suite:    th.Suite,
		UserId:   th.BasicUser.Id,
	}
	wc.SetConnectionID(model.NewId())
	wc.SetSession(&model.Session{})
	require.NoError(t, connIndex.Add(wc))

	wc.UserId = th.BasicUser2.Id
	wc.SetConnectionID(model.NewId())
	wc.SetSession(&model.Session{})
	require.NoError(t, connIndex.Add(wc))

	wc.UserId = specialUserID
	wc.SetConnectionID(model.NewId())
	wc.SetSession(&model.Session{})
	require.NoError(t, connIndex.Add(wc))

	now := time.Now()
	connIndex.Remove(wc)
	t.Logf("Time taken-----------: %s\n", time.Since(now))
}
```

cpoile reviewed (commented):
I'm really hesitant about using slice indexes like this.  It's error prone and confusing.
This might be crazy, but what about this: Since the index is essentially acting like a
manual map, turn the slices into maps using the `*WebConn` address as the index.
So, the new hubConnectionIndex would look like:

```
type hubConnectionIndex struct {
   // byUserId stores the connections for a given userID
   byUserId map[string]map[*WebConn]*WebConn
   // byChannelID stores the connections for a given channelID.
   byChannelID map[string]map[*WebConn]*WebConn
   // byWebConn is the reverse of byChannelID: it stores the list of channelIds by *WebConn 
   byWebConn map[*WebConn][]string
   byConnectionId map[string]*WebConn
    ...
 }
```

And then I think we don't need the connMetadata.
But we would need `byWebConn` (maybe named `channelIDsByWebConn` for clarity?) to store the channelIDs by `*WebConn`.
That would remove the slice index tracking, which as I mentioned is dangerous and confusing.
As for the `O(m*n)` building up the cache, I don't think we can avoid that, given the requirements...
I think this might simplify things, but I haven't written out the code so I might be missing a side effect.

agnivade commented:
@saturninoabril - I am seeing a

```
Body: {
  "id": "api.license.request_trial_license.app_error",
  "message": "Unable to get a trial license, please try again or contact with support@mattermost.com.",
  "detailed_error": "",
  "request_id": "jxpy46tnzjf33nnrgepzoek5xy",
  "status_code": 500
}
```

error in the smoke tests. Is there anything I can do to resolve this?

agnivade commented:
@cpoile - Nice! That's a neat suggestion. I implemented it half-way, but then ran into some other tradeoffs. There are 2 things here:
1. We cannot do this for the channelId map. Because while removing, you have the userId in the webConn, so you can easily get wc.UserId to get the map entry and remove the webConn from that map. But for channels, we don't have that inside the webConn struct. So either we have to make a DB query again to get the channnelIDs for that user, or iterate the full set of channels in `i.byChannelID`. Both solutions don't look too good for me.
2. For the userId map, yes that's doable. However, there's one slight snag - which is in the `ForUser` method. Currently, we store the conns in a slice and we have an optimization where if the number of conns is 1, then we don't need to copy the slice. But when we store the elements in a map, we _always_ have to iterate the map and allocate a new slice every single time. But all hope is not lost here. From 1.23 onwards, there's a maps.Keys [method](https://pkg.go.dev/maps#Keys) which introduces the iterator pattern by which we can cleanly iterate the map keys without having to re-do much of the code. And since map deletion is allowed while iteration, we don't need to do any copies as well!

So I'd suggest to wait till we start using 1.23 (which should be soon because 1.24 is just out), and apply optimization #2. WDYT?

cpoile commented:
> 1. We cannot do this for the channelId map. Because while removing, you have the userId in the webConn, so you can easily get wc.UserId to get the map entry and remove the webConn from that map. But for channels, we don't have that inside the webConn struct. So either we have to make a DB query again to get the channnelIDs for that user, or iterate the full set of channels in `i.byChannelID`. Both solutions don't look too good for me.

Ok, but unless I'm missing something, I think my suggestion above addresses your concerns. First, we get rid of the `webConn` struct, and second:

```go
// byWebConn is the reverse of byChannelID: it stores the list of channelIds by *WebConn
byWebConn map[*WebConn][]string
```

> 2. For the userId map, yes that's doable. However, there's one slight snag - which is in the `ForUser` method. Currently, we store the conns in a slice and we have an optimization where if the number of conns is 1, then we don't need to copy the slice. But when we store the elements in a map, we _always_ have to iterate the map and allocate a new slice every single time. But all hope is not lost here. From 1.23 onwards, there's a maps.Keys [method](https://pkg.go.dev/maps#Keys) which introduces the iterator pattern by which we can cleanly iterate the map keys without having to re-do much of the code. And since map deletion is allowed while iteration, we don't need to do any copies as well!
> 
> So I'd suggest to wait till we start using 1.23 (which should be soon because 1.24 is just out), and apply optimization #2. WDYT?

I think... IMO the optimization (for a slice of length 1) is nice, but the benefits of not having to track slice positions and all the trickery and trouble that brings is not worth it. Plus, we get the 1.23 optimization shortly anyway.

agnivade commented:
> Ok, but unless I'm missing something, I think my suggestion above addresses your concerns. 

Ahh sorry I missed that. Sorry yes, let me take a look at that.

agarciamontoro reviewed (commented):
Thank you for working on this, @agnivade! As discussed elsewhere, it's quite a complex piece of code, so kudos for that.

Some comments:

1. I agree with @cpoile that the slice indexes tricks, although clever, are a bit confusing and possibly error-prone. Looking forward to the map implementation, if it turns out to be possible.
2. The idea of the optimization seems sensible to me in general. Could you remind me why we're hiding it behind the flag? You said that this will still cause issues when "a large number of users suddenly join/leave a channel". Could you expand on this?
3. This is not from this PR, and not something to fix here, but as I'm trying to understand the whole context, here it goes: I find it quite confusing that the `ShouldSendEvent` function relies on something else happening in web_hub that is now conditional to a config knob. I don't have any solution, though just complaints :joy: 
4. Bonus question 1: do you have some gut feeling on why this was not caught by the performance comparisons? Is there something we're not stressing that we should?
5. Bonus question 2: would it be possible to have a (synchronous, if possible) knowledge sharing session at some point to go over the whole hub code, because there's so many moving (and critical) parts here that I feel I may be missing important context. I've spent quite a lot of time today understanding everything, but I'd like to learn about how you see it as well :)

Again, thanks for the work on this!

agnivade commented:
@cpoile / @agarciamontoro - PTAL

agnivade commented on server/channels/app/platform/web_hub.go:
```diff
@@ -828,19 +834,15 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
+		for _, chID := range connMeta.channelIDs {
+			if channelConns, ok := i.byChannelID[chID]; ok {
+				delete(channelConns, wc)
+				if len(channelConns) == 0 {
+					delete(i.byChannelID, chID)
```
Good point. There's no perf tradeoff here :) So I wil go ahead and remove this for simplicity.

agarciamontoro reviewed (changes requested):
Looks great, thank you :) I've left just a question below on some code that may be redundant.

agnivade commented on server/channels/app/platform/web_hub.go:
```diff
@@ -733,97 +739,116 @@ func closeAndRemoveConn(connIndex *hubConnectionIndex, conn *WebConn) {
+		if len(userConns) == 0 {
+			delete(i.byUserId, wc.UserId)
+		} else {
+			i.byUserId[wc.UserId] = userConns
```
Somewhere in the past I thought there was some assignment limitation in Go where maps of maps could not be accessed in a single line, and that you have to break the internal map into a new assignment. Clearly I was misremembering with something else. I just tested it, and it's unnecessary. Thanks.

agnivade commented on server/channels/app/platform/web_hub.go:
```diff
@@ -862,26 +907,31 @@ func (i *hubConnectionIndex) Has(wc *WebConn) bool {
+	if !ok {
+		return nil
+	}
+
+	// Move to using maps.Keys to use the iterator pattern with 1.23.
+	// This saves the additional slice copy.
```
Done https://mattermost.atlassian.net/browse/MM-63130

agnivade commented:
@agarciamontoro - PTAL

cpoile commented on server/channels/app/platform/web_hub.go:
```diff
@@ -740,7 +740,7 @@ func closeAndRemoveConn(connIndex *hubConnectionIndex, conn *WebConn) {
 type connMetadata struct {
-	channelIndices map[string]int
+	channelIDs []string
 }
```
Maybe a type alias would be more appropriate then?  (we don't need the struct)

agnivade replied:
Hmm .. what would you name the type? `byWebConnWithChannelIds`?

cpoile replied:
If you don't want to change the pattern, you could do:
```go
type channelIDs = []string  // this replaces the connMetadata struct
...
byConnection map[*WebConn]channelIDs
```

Or if you want the name to be more descriptive:
```go
type channelIDs = []string  // this replaces the connMetadata struct
...
channelIDsByConnection map[*WebConn]channelIDs
```

Something like that?

agnivade replied:
sure, will change this as part of https://mattermost.atlassian.net/browse/MM-63130.

agarciamontoro reviewed (approved):
Nice, much simpler! Thanks, Agniva :)

DHaussermann reviewed (approved):
Tested and passed
I setup local HA environment so I could be certain I am seeing WS events travel across to a separate cluster instance. And I have control over the `EnableWebHubChannelIteration` flag locally

- Testes that channel add/remove WS events for channel are specific to channel members as well as post and typing events within the channel
- Tested add/remove and other activity to ensure that Team scope is respected
- Used DM channel typing and posts to test user specific WS event scope is respected 
- Tested Flaky connection to ensure WS event function normally once connectivity is restored

- Set `EnableWebHubChannelIteration` to true and repeated testing
- Setup a `/cloud` server in HA and repeated testing in case the infrastructure there surfaces a concern I could not see locally

No issues found.

Thanks @agnivade. 
LGTM!

mattermost-build commented:
Cherry pick is scheduled.

agnivade commented:
@cwarnermm - Another FYI that this PR adds a new config setting. It's documented in the release notes.

amyblais commented:
/cherry-pick release-10.6

mattermost-build commented:
Cherry pick is scheduled.

amyblais commented:
/cherry-pick release-10.4

mattermost-build commented:
Cherry pick is scheduled.

amyblais commented:
/cherry-pick release-10.3

mattermost-build commented:
Cherry pick is scheduled.

mattermost-build commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-mattermost-#30178-upstream-release-10.4-1739517486
Switched to a new branch 'automated-cherry-pick-of-mattermost-#30178-upstream-release-10.4-1739517486'
Branch 'automated-cherry-pick-of-mattermost-#30178-upstream-release-10.4-1739517486' set up to track remote branch 'release-10.4' from 'upstream'.

+++ About to attempt cherry pick of PR #30178 with merge commit da7192246ea6e30bbf01a351f3e410230e6cc54c.

Auto-merging server/channels/api4/post_test.go
CONFLICT (content): Merge conflict in server/channels/api4/post_test.go
Auto-merging server/channels/app/platform/web_conn.go
Auto-merging server/channels/app/platform/web_hub.go
Auto-merging server/public/model/config.go
error: could not apply da7192246e... MM-62960: Improve webConn remove performance from hubConnectionIndex (#30178)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

+++ Conflicts detected:

UU server/channels/api4/post_test.go
Aborting.

+++ Aborting in-progress git cherry-pick.

+++ Returning you to the master branch and cleaning up.

```

amyblais commented:
@agnivade Can you help open a manual cherry-pick PR for ``release-10.4`` branch?
